File: medicompare/frontend/js/comparison.js
/* =========================================================
   MediCompare — comparison.js
   ========================================================= */

Auth.requireAuth();

const params = new URLSearchParams(window.location.search);
const prescriptionId = params.get('id');

document.getElementById('back-link').href = prescriptionId ? `analysis.html?id=${prescriptionId}` : 'scan.html';

if (!prescriptionId) {
  document.getElementById('comparisons-mount').innerHTML =
    `<div class="card state-block"><h3>No prescription selected</h3><p>Go back and scan a prescription first.</p></div>`;
} else {
  loadComparisons();
}

async function loadComparisons() {
  try {
    const prescription = await apiRequest(`/api/prescriptions/${prescriptionId}`);
    document.getElementById('view-history-link').href = `history.html?patient_id=${prescription.patient_id}`;

    if (prescription.medicines.length === 0) {
      document.getElementById('comparisons-mount').innerHTML =
        `<div class="card state-block"><h3>No medicines to compare</h3><p>Go back and add at least one medicine.</p></div>`;
      return;
    }

    const results = [];
    for (const med of prescription.medicines) {
      try {
        const result = await apiRequest('/api/comparisons', {
          method: 'POST',
          body: { prescription_medicine_id: med.id },
        });
        results.push(result);
      } catch (err) {
        results.push({
          prescription_medicine_id: med.id,
          prescribed_name: med.medicine_name,
          match_status: 'unmatched',
          options: [],
          prescribed_price: null,
          lowest_generic_price: null,
          lowest_jan_aushadhi_price: null,
          potential_saving: 0,
        });
      }
    }

    renderSummary(results);
    renderComparisons(results);
  } catch (err) {
    document.getElementById('comparisons-mount').innerHTML =
      `<div class="card state-block"><h3>Couldn't load comparisons</h3><p>${err.message}</p></div>`;
  }
}

function renderSummary(results) {
  const matched = results.filter((r) => r.match_status === 'matched');
  const totalPrescribed = matched.reduce((sum, r) => sum + (r.prescribed_price || 0), 0);
  const totalSaving = matched.reduce((sum, r) => sum + (r.potential_saving || 0), 0);
  const totalLowest = totalPrescribed - totalSaving;

  const mount = document.getElementById('summary-mount');
  if (matched.length === 0) {
    mount.innerHTML = '';
    return;
  }

  mount.innerHTML = `
    <div class="savings-summary">
      <div>
        <div class="savings-summary-label">Total potential saving across this prescription</div>
        <div class="price price-xl">${formatCurrency(totalSaving)}</div>
      </div>
      <div class="savings-summary-breakdown">
        <div><span>Prescribed total</span><span class="price">${formatCurrency(totalPrescribed)}</span></div>
        <div><span>Lowest compared</span><span class="price">${formatCurrency(totalLowest)}</span></div>
      </div>
    </div>`;
}

function renderComparisons(results) {
  const mount = document.getElementById('comparisons-mount');
  mount.innerHTML = results.map((r) => comparisonBlockHtml(r)).join('');
}

function comparisonBlockHtml(r) {
  if (r.match_status !== 'matched') {
    return `
      <div class="card medicine-comparison-block">
        <div class="mc-head"><div><h3>${escapeHtml(r.prescribed_name)}</h3></div></div>
        <div class="unmatched-block">
          <p>We couldn't find a verified catalogue match for this medicine yet, so no comparison is available.</p>
          <span class="field-hint">You can still discuss generic options with your pharmacist directly.</span>
        </div>
      </div>`;
  }

  const prescribedOpt = r.options.find((o) => o.type === 'Prescribed');
  const genericOpts = r.options.filter((o) => o.type === 'Generic');
  const jaOpts = r.options.filter((o) => o.type === 'Jan Aushadhi');
  const bestGeneric = genericOpts.length ? genericOpts.reduce((a, b) => (a.price < b.price ? a : b)) : null;
  const bestJA = jaOpts.length ? jaOpts.reduce((a, b) => (a.price < b.price ? a : b)) : null;

  const maxPrice = Math.max(r.prescribed_price || 0, 1);

  const ladderRow = (label, price, colorVar) => {
    const width = price ? Math.max(6, Math.round((price / maxPrice) * 100)) : 0;
    return `
      <div class="ladder-row">
        <span class="ladder-label">${label}</span>
        <span class="ladder-track"><span class="ladder-fill" style="width:${width}%; background:${colorVar}"></span></span>
        <span class="price">${price !== null && price !== undefined ? formatCurrency(price) : '—'}</span>
      </div>`;
  };

  const optionCard = (opt, typeClass, tagText) => {
    if (!opt) return '';
    const isBest = tagText ? `<span class="option-save-tag">${tagText}</span>` : '';
    return `
      <div class="option-card ${typeClass}">
        ${isBest}
        <div class="option-type-label">${opt.type}</div>
        <div class="option-brand">${escapeHtml(opt.brand)}</div>
        <div class="option-meta">${escapeHtml(opt.manufacturer || '')} · ${escapeHtml(opt.pack_size || '')}</div>
        <div class="price">${formatCurrency(opt.price)}</div>
      </div>`;
  };

  const savingTag = r.potential_saving > 0 ? `Save ${formatCurrency(r.potential_saving)}` : null;

  const tableRows = r.options.map((o) => `
    <tr>
      <td data-label="Option">${o.type}</td>
      <td data-label="Brand">${escapeHtml(o.brand)}</td>
      <td data-label="Salt">${escapeHtml(o.salt)}</td>
      <td data-label="Strength">${escapeHtml(o.strength)}</td>
      <td data-label="Manufacturer">${escapeHtml(o.manufacturer || '—')}</td>
      <td data-label="Pack size">${escapeHtml(o.pack_size || '—')}</td>
      <td data-label="Price" class="price">${formatCurrency(o.price)}</td>
      <td data-label="Source"><span class="field-hint">${escapeHtml(o.source)}</span></td>
    </tr>`).join('');

  return `
    <div class="card medicine-comparison-block">
      <div class="mc-head">
        <div>
          <h3>${escapeHtml(r.prescribed_name)}</h3>
          <span class="mc-sub">Same composition matches only — active ingredient, strength & dosage form</span>
        </div>
        ${r.potential_saving > 0 ? `<span class="badge badge-medium"><span class="badge-dot"></span>Save ${formatCurrency(r.potential_saving)}</span>` : `<span class="badge badge-neutral">Already lowest priced</span>`}
      </div>

      <div class="ladder-demo" style="padding:0; margin-bottom: var(--space-5);">
        ${ladderRow('Prescribed', r.prescribed_price, 'var(--color-ink-faint)')}
        ${r.lowest_generic_price !== null ? ladderRow('Generic', r.lowest_generic_price, 'var(--color-primary)') : ''}
        ${r.lowest_jan_aushadhi_price !== null ? ladderRow('Jan Aushadhi', r.lowest_jan_aushadhi_price, 'var(--color-accent)') : ''}
      </div>

      <div class="comparison-columns">
        ${optionCard(prescribedOpt, 'type-prescribed', null)}
        ${optionCard(bestGeneric, 'type-generic', bestGeneric && bestGeneric.price < (r.prescribed_price || 0) ? `Save ${formatCurrency(r.prescribed_price - bestGeneric.price)}` : null)}
        ${optionCard(bestJA, 'type-jan-aushadhi', bestJA && bestJA.price < (r.prescribed_price || 0) ? `Save ${formatCurrency(r.prescribed_price - bestJA.price)}` : null)}
      </div>

      <details>
        <summary class="field-hint" style="cursor:pointer; margin-bottom: var(--space-3);">View all ${r.options.length} options in detail</summary>
        <div class="comparison-table-wrap">
          <table class="comparison-table responsive-table">
            <thead>
              <tr><th>Option</th><th>Brand</th><th>Salt</th><th>Strength</th><th>Manufacturer</th><th>Pack size</th><th>Price</th><th>Source</th></tr>
            </thead>
            <tbody>${tableRows}</tbody>
          </table>
        </div>
      </details>

      <div class="substitution-note">
        Same active ingredient, strength, and dosage form as prescribed. Please confirm any switch with your doctor or pharmacist.
      </div>
    </div>
  `;
}

function escapeHtml(str) {
  const div = document.createElement('div');
  div.textContent = str || '';
  return div.innerHTML;
}

File: medicompare/backend/app/services/comparison.py
"""
Builds a price comparison result for a single prescription medicine, using
only deterministic "same composition" matches from services/matching.py.
"""
from typing import Optional
from sqlalchemy.orm import Session

from app.models import Medicine, PrescriptionMedicine
from app.services.matching import find_best_catalogue_match, find_same_composition


def build_comparison(db: Session, pm: PrescriptionMedicine) -> dict:
    matched: Optional[Medicine] = None
    if pm.matched_medicine_id:
        matched = db.query(Medicine).filter(Medicine.id == pm.matched_medicine_id).first()
    if matched is None:
        matched = find_best_catalogue_match(db, pm.medicine_name, pm.strength or "")

    options = []
    prescribed_price = None
    lowest_generic = None
    lowest_jan_aushadhi = None

    if matched is None:
        return {
            "prescription_medicine_id": pm.id,
            "prescribed_name": pm.medicine_name,
            "match_status": "unmatched",
            "options": [],
            "prescribed_price": None,
            "lowest_generic_price": None,
            "lowest_jan_aushadhi_price": None,
            "potential_saving": 0.0,
        }

    prescribed_price = matched.price
    options.append({
        "type": "Prescribed",
        "medicine_id": matched.id,
        "brand": matched.brand_name,
        "salt": matched.active_ingredient,
        "strength": matched.strength,
        "manufacturer": matched.manufacturer,
        "pack_size": matched.pack_size,
        "price": matched.price,
        "source": matched.source,
    })

    alternatives = find_same_composition(db, matched)
    for alt in alternatives:
        is_generic = alt.generic_available or alt.brand_name.lower() == alt.generic_name.lower()
        opt_type = "Generic" if is_generic else "Generic"
        options.append({
            "type": opt_type,
            "medicine_id": alt.id,
            "brand": alt.brand_name,
            "salt": alt.active_ingredient,
            "strength": alt.strength,
            "manufacturer": alt.manufacturer,
            "pack_size": alt.pack_size,
            "price": alt.price,
            "source": alt.source,
        })
        if lowest_generic is None or alt.price < lowest_generic:
            lowest_generic = alt.price

        if alt.jan_aushadhi_available and alt.jan_aushadhi_price:
            options.append({
                "type": "Jan Aushadhi",
                "medicine_id": alt.id,
                "brand": alt.brand_name,
                "salt": alt.active_ingredient,
                "strength": alt.strength,
                "manufacturer": "Jan Aushadhi Kendra",
                "pack_size": alt.pack_size,
                "price": alt.jan_aushadhi_price,
                "source": "PMBJP demo reference",
            })
            if lowest_jan_aushadhi is None or alt.jan_aushadhi_price < lowest_jan_aushadhi:
                lowest_jan_aushadhi = alt.jan_aushadhi_price

    # The prescribed medicine itself might have a Jan Aushadhi price
    if matched.jan_aushadhi_available and matched.jan_aushadhi_price:
        if lowest_jan_aushadhi is None or matched.jan_aushadhi_price < lowest_jan_aushadhi:
            lowest_jan_aushadhi = matched.jan_aushadhi_price

    candidate_lows = [p for p in [lowest_generic, lowest_jan_aushadhi] if p is not None]
    lowest_overall = min(candidate_lows) if candidate_lows else prescribed_price
    potential_saving = max(0.0, round(prescribed_price - lowest_overall, 2))

    return {
        "prescription_medicine_id": pm.id,
        "prescribed_name": pm.medicine_name,
        "match_status": "matched",
        "options": options,
        "prescribed_price": prescribed_price,
        "lowest_generic_price": lowest_generic,
        "lowest_jan_aushadhi_price": lowest_jan_aushadhi,
        "potential_saving": potential_saving,
    }
